mlir/include/mlir/Dialect/OpenMP/OpenMPOps.td

```diff
@@ -618,6 +618,22 @@ def AtomicUpdateOp : OpenMP_Op<"atomic.update"> {
   let verifier = [{ return verifyAtomicUpdateOp(*this); }];
 }
 
+//===----------------------------------------------------------------------===//
+// [5.1] 2.21.2 threadprivate Directive
+//===----------------------------------------------------------------------===//
+
+def ThreadprivateOp : OpenMP_Op<"threadprivate"> {
+  let summary = "threadprivate directive";
+  let description = [{
+    The threadprivate directive specifies that variables are replicated, with
```

Comment:
Can we please have a little more elaborate description here - why this was necessary and why is this not added as a part of the construct itself - as a clause?

Reply:
Having the mangled name as an attribute and then comparing based on name doesnt look like the optimal way to do this. Can we instead compare this on the basis of the global itself? So, how about the following syntax - 
```
llvm.mlir.global internal @_QFsubEx() : i32 {
  %0 = llvm.mlir.undef : i32
  llvm.return %0 : i32
}
llvm.func sample(%arg1: i32, %arg2: i32) -> () {
  %x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>
  llvm.store %arg1, %x : !llvm.ptr<i32>
  omp.parallel {
    %y = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>
    llvm.store %arg2, %y : !llvm.ptr<i32>
  }
}
```
Let me know if this doesn't work.

Reply:
> Can we please have a little more elaborate description here - why this was necessary and why is this not added as a part of the construct itself - as a clause?

Not sure what more do you want. The threadprivate directive itself is not one clause. It is one declarative directive.

Something like  `%x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>` won't work. Please check the lowering of the common block element.

Reply:
> Not sure what more do you want. The threadprivate directive itself is not one clause. It is one declarative directive.
I somehow missed the next paragraph. It looks good. Please ignore this comment.
> Something like %x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32> won't work. Please check the lowering of the common block element.
You mean the PFT to FIR lowering? I am worried that there is some redundancy in the sym_addr and sym_name. Can you please clarify the issues with this. I was proposing the following - 
```
!OMPDialect: %x = omp.threadprivate @_QFEi
!LLVM-IR: %x = call i8* @__kmpc_threadprivate_cached(%struct.ident_t* @[[GLOB2]], i32 [[THREAD2]], i8* bitcast (i32* @_QFEi to i8*), i64 4, i8*** @_QFEi.cache)
!$omp threadprivate(i)

```

Reply:
> > Not sure what more do you want. The threadprivate directive itself is not one clause. It is one declarative directive.
> 
> I somehow missed the next paragraph. It looks good. Please ignore this comment.
> 
> > Something like %x = omp.threadprivate @_QFsubEx : !llvm.ptr won't work. Please check the lowering of the common block element.
> 
> You mean the PFT to FIR lowering? I am worried that there is some redundancy in the sym_addr and sym_name. Can you please clarify the issues with this. I was proposing the following -
> 
> ```
> !OMPDialect: %x = omp.threadprivate @_QFEi
> !LLVM-IR: %x = call i8* @__kmpc_threadprivate_cached(%struct.ident_t* @[[GLOB2]], i32 [[THREAD2]], i8* bitcast (i32* @_QFEi to i8*), i64 4, i8*** @_QFEi.cache)
> !$omp threadprivate(i)
> ```

Yes. The threadprivate directive is kind of adding one attribute into the variables. `clang` does add the attribute into the symbol directly. However, it is not allowed in FIR. For symbols in common block, only one threadprivate operation is added into the common block, not its memebers. This means we add the attribute to the whole common block. `%x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>` changes the meaning. But it seems ok to generate the correct LLVM IR. For variables using in host-association, there is no this mangled name currently. For polymorphics and data in future C++ IR, I am not sure if it can use one single addressof operation to get the the value of symbol. With these concerns, I prefer to not change the original operations. Instead, only add one operation, which is more extendable. What do you think?

Regarding to the redundancy concern of the sym_addr and sym_name, sym_addr does store the info of sym_name. However, getting the sym_name from sym_addr is not supported in LLVMIR Dialect. In FIR, `sym_name` is `uniq_name` in GlobalOp/AllocaOp. In other IR, it may be others. Check this [comment](https://github.com/flang-compiler/f18-llvm-project/pull/1135#discussion_r831129255).

Reply:
For the next few lines, "proposed syntax" means the syntax `%x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>` and the "current implementation" means what the patch implements.

> This means we add the attribute to the whole common block.

For common block `/blk/` the global symbol is @_QBblk. Are you saying that with the proposed syntax we cannot handle common block semantics properly? Why would `%x = omp.threadprivate @_QBblk : !llvm.ptr<i32>` not work? I am sorry but I am missing something. Can you please give me the problematic Fortran code where we wont be able to express threadprivate using the proposed syntax in FIR but the current implementation can express it properly? 

Also, as you have concern with the PFT to MLIR stage, I am assuming that for the proposed syntax there don't seem to be any obvious issues from MLIR to LLVM IR.

Reply:
> For the next few lines, "proposed syntax" means the syntax %x = omp.threadprivate @_QFsubEx : !llvm.ptr<i32>.

The comments do not mean addressing the symbol. It means adding something to the addressing of symbols. As you can see, `sym_addr` refers to the addressing of the symbol and `sym_name` refers to adding the mangled symbol name.

> For common block `/blk/` the global symbol is @_QBblk. Are you saying that with the proposed syntax we cannot handle common block semantics properly? Why would `%x = omp.threadprivate @_QBblk : !llvm.ptr<i32>` not work? I am sorry but I am missing something. Can you please give me the problematic Fortran code where we wont be able to express threadprivate using the proposed syntax in FIR but the current implementation can express it properly?

`%x = omp.threadprivate @_QBblk : !llvm.ptr<i32>` means addressing of common block. The threadprivate private itself does not have this meaning. With the current implementation, we are creating the operation for common block members. The operation is attched in addressing of common block to add something (here means threadprivate info) for it. The data-sharing attributes to common block also means each common block member has this attribute.

> Also, as you have concern with the PFT to MLIR stage, I am assuming that for the proposed syntax there don't seem to be any obvious issues from MLIR to LLVM IR.

Check this example:
```
program main
  integer, save :: a
  !$omp threadprivate(a)
  call sub()
contains
  subroutine sub()
    a = 2
  end
end
```
```
$ bbc -emit-fir test.f90 && cat test.mlir
func @_QFPsub(%arg0: !fir.ref<tuple<!fir.ref<i32>>> {fir.host_assoc}) {
  %c0_i32 = arith.constant 0 : i32
  %0 = fir.coordinate_of %arg0, %c0_i32 : (!fir.ref<tuple<!fir.ref<i32>>>, i32) -> !fir.llvm_ptr<!fir.ref<i32>>
  %1 = fir.load %0 : !fir.llvm_ptr<!fir.ref<i32>>
  %c2_i32 = arith.constant 2 : i32
  fir.store %c2_i32 to %1 : !fir.ref<i32>
  return
}
```
`%x = omp.threadprivate @mangled_name` won't work since there is no mangled name. I have no idea currently how to handle this case. I think the current implementation has more potential to handle this case than `%x = omp.threadprivate @mangled_name`. In the current implementation, we can add more arguments, maybe the host-associated addressing (`%1 = fir.load %0`).

> I agree with this point. Does the proposed syntax force any extensions to original operations?

That's what I think it should be. With more data type and data usage cases, this operation may should be extended. Currently, I mean host-assocation and polymorphics.

Reply:
> Check this example:

Okay, this example makes it clear. There is a global `@_QFEa` but it is loaded and passed as an argument to the `_QFPsub` function. I think the bigger question here is why the global is going in as an argument when it should actually be loaded like a global. I am not sure about the implications in other parts of the language but I think the solution to this should be to load the global exactly like it gets loaded in main
```
%0 = fir.address_of(@_QFEa) : !fir.ref<i32>
```
Once we always have the access to the global, would it make this implementation simpler? Please correct me if my interpretation of the current flow is incorrect - we are going from global (in Fortran) to argument (in FIR) and then back to global (in LLVM IR) - this flow doesn't seem right to me and this flow forces us to have both sym_name and sym_addr. @kiranchandramohan @PeixinQiao @schweitzpgi @clementval is loading the global in FIR instead of passing it as an argument to the function feasible? And would it affect the other parts of the language adversely?

Assuming that we must load from a value and loading from the global is not an option, `%x = omp.threadprivate @mangled_name` doesn't work. Thank you for the explanation @PeixinQiao. This also clears explains the redundancy.

Reply:
I haven't gone through this in detail. But can we have something like the critical construct, where there is a declaration and subsequent uses just reference the declaration?

Reply:
> Once we always have the access to the global, would it make this implementation simpler?

That would solves the problem of host-assocation. But I don't think the current implementation of host-association can be changed. Please check the following example, the "a" is not always global.
```
program main
  integer :: a
  call sub()
contains
  subroutine sub()
    a = 2
  end
end
```

Reply:
> But I don't think the current implementation of host-association can be changed

This is what I wanted to understand. I will look up host-association to understand this better, but why not? Is it because it is a bigger change, or are there fundamental limitations to doing this?

Reply:
`gfortran` is wrong for this case. This `a` has implicit SAVE attribute, which is allowed in threadprivate directive in OpenMP standard.

What I would like you to check is not OpenMP code. Please check the FIR generated for the above case. The FIR for `a` is alloca op, which must be passed as argument in host-association.

Reply:
What I am proposing is that global variables must be accessed like global variables and normal member variables must be accessed like normal ones.
```
program main
  integer :: a
  integer, save :: b
  call sub()
contains
  subroutine sub()
    a = 2
    b = 4
  end
end
```
This example would pass `a` as an argument and `b` will be accessed like a global - which it should be.

Reply:
> gfortran is wrong for this case. This a has implicit SAVE attribute, which is allowed in threadprivate directive in OpenMP standard.

Hmm, I did not expect this. I figured the implicit attribute was only when the variables were initialized while declaring them. I will check this once again.

Reply:
> But can we have something like the critical construct, where there is a declaration and subsequent uses just reference the declaration?

@kiranchandramohan  Do you mean something like the following?
```
omp.threadprivate.declare @_QFEx1
%1 = fir.address_of(@_QFEx1) : !fir.ref<i32>
omp.threadprivate(%1 : !fir.ref<i32>)
```

Reply:
> This example would pass `a` as an argument and `b` will be accessed like a global - which it should be.

This also sounds reasonable to me. But I am not sure if FIR team has some special reason to do this or if they are willing to change the current implemenation for host-association.

Reply:
Even if the value for variable `b` is changed into global Op. There is still the problem for variable `a` in threadprivate directive. Some previous discussions are [here](https://github.com/flang-compiler/f18-llvm-project/issues/1093).

Reply:
Okay, I see. If we can fix the previous issue #1093, we will be able to have a much simpler design for `threadprivate`. The issue is simple, but it is causing ripple effects in the OpenMPDialect. Right now, I am okay with both - the current implementation and Kiran's suggestion as I don't want to block this because of the FIR issue.

Reply:
I go back to check clang again and remember why I want to add one attribute for implementing the threadprivate operation. Please check this previous comment https://github.com/flang-compiler/f18-llvm-project/pull/1135#discussion_r732368811. Clang takes the `threadprivate` as one attribute of the symbols. There are two methods in clang to implement the threadprivate directive. Current classic-flang only supports the one as this patch. OMPIRBuilder has supported this runtime call. For the other method, clang adds `thread_local` to GlobalOp with some more changes and does not change the addressing operation. `%x = omp.threadprivate @_QFsubEx` deleted the original addressing operation, so it is not extendable for that method. Currently, I am not sure if adding `thread_local` can support all data types and all data usage cases in fortran, but making the ThreadprivateOp more extendable is a good way.